### DISCUSSION TITLE
PeriodicTimer with TimeProvider assertion is not stable in .net 8

### DISCUSSION THREAD

**Discussion by fingers10** [dotnet/aspnetcore#53922] [Q&A]
I have a simple `BackgroundService` in my asp.net 8 web api which performs an hourly backup using an `PeriodicTimer`. Here is the code sample,

```cs
public class OfflineHourlyDatabaseBackup(
	TimeProvider timeProvider,
	IServiceProvider serviceProvider) : BackgroundService
{
	protected override async Task ExecuteAsync(CancellationToken stoppingToken)
	{
		var timer = new PeriodicTimer(TimeSpan.FromHours(1), timeProvider);

		while (await timer.WaitForNextTickAsync(stoppingToken) && !stoppingToken.IsCancellationRequested) 
		{
			Backup();
		}
	}

	private void Backup()
	{
		using var scope = serviceProvider.CreateScope();
		var offlineDatabaseBackupService = scope.ServiceProvider.GetRequiredService<IOfflineDatabaseBackupService>();
		offlineDatabaseBackupService.Backup();
	}
}
```

Now I have added unit test to assert the hourly backup behaviour as shown below,

```cs
[Fact]
public async Task OfflineHourlyDatabaseBackup_ExecuteAsync()
{
    // Arrange
    var serviceCollection = new ServiceCollection();
    var mockOfflineDatabaseBackupService = new Mock<IOfflineDatabaseBackupService>();
    serviceCollection.AddSingleton(_ => mockOfflineDatabaseBackupService.Object);
    var timeProvider = new FakeTimeProvider();
    var serviceProvider = serviceCollection.BuildServiceProvider();

    // Act
    using var offlineHourlyDatabaseBackup = new OfflineHourlyDatabaseBackup(timeProvider, serviceProvider);
    await offlineHourlyDatabaseBackup.StartAsync(default);
    timeProvider.Advance(TimeSpan.FromHours(1));
    timeProvider.Advance(TimeSpan.FromHours(1));
    await offlineHourlyDatabaseBackup.ExecuteTask!;

    // Assert
    offlineHourlyDatabaseBackup.ExecuteTask.IsCompletedSuccessfully.Should().BeTrue();
    mockOfflineDatabaseBackupService.Verify(x => x.Backup(), Times.AtLeast(2));
}
```

The test assertion is not stable. It passes sometimes and fails sometimes with the following error message

> Moq.MockException :  Expected invocation on the mock at least 2 times,
> but was 1 times: x => x.Backup()

Please can you help me understand what I'm doing wrong here?

**Comment by BrennanConroy** [comment]
Doesn't that code hang forever on `await offlineHourlyDatabaseBackup.ExecuteTask!;`?

**Reply by fingers10** [reply]
@BrennanConroy yes it happens sometimes. Any inputs on how to resolve this? Please guide me on what I'm missing here..

**Comment by martincostello** [comment]
Don't you need to pass a CancellationToken to the service so that it will terminate the loop?

**Reply by fingers10** [reply]
@martincostello ExecuteTask doesn't have an option to supply `CancellationToken`

**Reply by martincostello** [reply]
No it doesn't, but the `CancellationToken` passed to `StartAsync()` passes it to the `ExecuteAsync()` method associated with that task: [code](https://github.com/dotnet/runtime/blob/35562ee5ac02c68d42d5b77fb0af09123d79c3ba/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs#L40-L46)

**Reply by fingers10** [reply]
I thought the `default` will take care of it in `await offlineHourlyDatabaseBackup.StartAsync(default);`. Let me explicitly create and pass and check and get back

**Reply by fingers10** [reply]
@martincostello no it doesnt work

**Reply by fingers10** [reply]
@martincostello any thoughts on this?
